Test/git-issues/git-issue-889a.dfy

```diff
@@ -0,0 +1,146 @@
+// RUN: %dafny "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// This file includes verification tests
+
+// ---------------------------------
+
+newtype int8 = x | -128 <= x < 128
+newtype smallneg = r | -1.0 <= r <= 0.0
+
+method GoodNewtypeLiterals() returns (b: int8, r: smallneg) {
+  if {
+    case true =>
+      b := -127 - 1;
+    case true =>
+      b := (-128) as int8;
+    case true =>
+      b := -128 as int8;  // `-` binds stronger than `as`, so this is the same as the previous line
+    case true =>
+      b := -128;
+    case true =>
+      b := - /* some space here */ 128;  // it's two tokens, but it gives one negative literal
+  }
+  assert b + 64 + 64 == 0;
+
+  if {
+    case true =>
+      r := -1.0 as smallneg;
+    case true =>
+      r := -1.0;
+  }
+}
+
+method BadNewtypeLiterals() returns (b: int8, r: smallneg) {
+  if
+  case true =>
+    b := -(128 as int8);  // error: 128 is not a int8
+  case true =>
+    b := -128;
+    assert 64 + 64 + b == 0;  // error: the subexpression "64 + 64" gives a value that's too large
+  case true =>
+    r := -(1.0);  // error: 1.0 is not a value of type smallneg
+}
+
+method MoreBadLiterals() returns (b: int8, r: smallneg) {
+  if
+  case true =>
+    b := --128;  // error: the result of the unary minus, 128, is not a int8
+  case true =>
+    b := - - 128;  // error: the result of the unary minus, 128, is not a int8
+  case true =>
+    r := --1.0;  // error: the result of the unary minus, 1.0, is not a value of type smallneg
+  case true =>
+    r := - - 1.0;  // error: the result of the unary minus, 1.0, is not a value of type smallneg
+}
+
```

Comment:
For 'good literals' is -- allowed?  e.g. for bv8 is --5 a literal, equal to 5, or is it a unary minus expression -(-5)?

Reply:
For an integer expression, `--5` is UnaryMinus(-5). For a bitvector, it is UnaryMinus(UnaryMinus(5)).

Reply:
I'm fine with this behavior. I think it is worth a little test to be sure this is the behavior. I believe the current grammar at least supports this, but I could imagine a grammar that said a literal is [0-9]+|[-]{literal} or something like that.

Reply:
I added a test to make sure that `--5` isn't in the future accidentally changed to stand for the literal `5` (instead of the unary-minus of `-5`).

I did this using a `newtype exactly5 = x | x == 5` rather than trying to use `--5` in a `case` label as we just discussed in person, because the latter gives a parse error.